src/event.rs

```diff
@@ -629,10 +629,18 @@ pub struct KeyboardInput {
 
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
-pub enum CompositionEvent {
-    CompositionStart(String),
-    CompositionUpdate(String, usize),
-    CompositionEnd(String),
+pub enum PreeditState {
```

Comment:
I think we can name it a bit differently, since it's not only about preedit state, since you have `commit`, etc. I'd also like to see adoption of 
```
pub enum CompositionEvent {
    CompositionEnabled,
    CompositionPreedit(String, Option<usize>, Option<usize>),
    CompositionCommit(String),
    CompositionDisabled,
}
```

That I've mentioned on #1497 if you don't have any objects, unfortunately I don't really know much about X11, so it's a bit hard for me to update its code myself.

Also we can name rename `PreeditState` to `IMEEvent`, and have

```rust
pub enum IMEEvent {
    Enabled,
    Preedit(String, Option<usize>, Option<usize>),
    Commit(String),
    Disabled,
}
```

So I guess Start and Update should be transformed into `Preedit`, and End should be transformed into Commit.

Enabled and Disabled should be implemented, and I guess you have that information from X server, at least I can see something like that in xim docs.

Reply:
I think you miss Start (or maybe PreeditStart?) event. Other than that, I think it looks fine.

Reply:
>I think you miss Start (or maybe PreeditStart?) event. Other than that, I think it looks fine.

Could you explain why `PreeditStart` is required? on Wayland for example everything starts with just `Preedit`.

Reply:
In Wep API, CompositionStart exist in the specification. And in x11, every IME session starts with preedit start event. So, I believe something to tell us when a new IME session starts is necessary.

Reply:
By start you mean that input method got open or just new input? Since Wayland has a concept of 'Enabled', when user uses its IME, like they opened 'fcitx' or 'ibus' and focused your app. But there's no notion of 'Start', and everything is communicated via 'Preedit' and 'Commit'.

My question was about start, since it looked like `Preedit("", None, None)` to me.

Reply:
My main concern is that I can't really send anything about on `PreeditStart` on Wayland, since I don't have a `Start` at all, I can synthesize a `PreeditStart("")`, but I don't feel like it'll make sense, so I was wondering  if we can send `Preedit("", None, None)` instead of start on those platforms.

Reply:
> By start you mean that input method got open or just new input?

I mean start a input method editor got open. For example, the user input "きょうは"(Enter)"いい"（Enter）"てんきですね"(Enter), then the start event will happen just before the user start input k of "きょうは", "i" of "いい", and "ｔ" of "てんきですね".
It is different concept from enable event. The enable event will happen when the user press the key to switch IME mode (e.g. "Zenkaku/Hankaku" button in JP keyboard).

> My main concern is that I can't really send anything about on PreeditStart on Wayland

In X11 and Web API allows to start with some texts. For example, I have a committed text "きょうは", and I want to convert it into Kanji after commit, then I can select this string and press key to convert, then the new session will start with "きょうは".
If you have Japanese IME, you  can see those behaviors in MDN doc page
https://developer.mozilla.org/en-US/docs/Web/API/Element/compositionstart_event

Reply:
>In X11 and Web API allows to start with some texts. For example, I have a committed text "きょうは", and I want to convert it into Kanji after commit, then I can select this string and press key to convert, then the new session will start with "きょうは".

Yeah, I'll just get `Preedit` for that on Wayland, not start, and my primary concern why can't we merge Start and `Preedit`, since it's literally doesn't matter what reports initial text, right? Or you want different handling for `Start("Hello")` and `Preedit("Hello", None, None)`, I'd assume for both cases you shouldn't insert text into widget, right? Since it's a job of `Commit`.

Reply:
> Or you want different handling for Start("Hello") and Preedit("Hello", None, None)

So, your suggestion is the application can detect the session start by checking the second argument is empty?
I feel this way is not intuitive.
I think start event in IME is common since this is defined not only in X11, also Web API and other famous framework, like gtk. So, I prefer to having a separate member to tell the session start.

Reply:
>So, your suggestion is the application can detect the session start by checking the second argument is empty?

But why do you need to detect session start at all? The start is when you've got any IME event.

I mean, I don't understand how end users will handle `Start` differently to `Preedit`? Could you give an example where handling of those 2 are different and it makes sense? I get that it's common API for X11 and Web, but it really feels like someone decided to add redundant event that you can mirror with other events.

Reply:
Hmm, I don't have an idea of specific usage for Start event for now, but Start event looks relatively common, so I'm afraid that there is some needs.
But if there is no special use case, it might be OK to combine them.

Reply:
> I mean, I don't understand how end users will handle Start differently to Preedit? Could you give an example where handling of those 2 are different and it makes sense?

A example I met before is, I need to know whether I should handle `ReceiveCharacter` or not. When I get `PreeditStart`, I can set a flag like `isPreediting`, if `isPreediting` is true, I will ignore `ReceiveCharacter` event.

(However, we can do it in winit though, when preediting, we don't trigger `ReceiveCharacter`.)

I think it may be useful when user need to do stuff depend on IME state.

Reply:
If we do preedit we don't need to send receivedchar, since the data is being send via preedit.

Reply:
I'm not sure, but there may be other events in the similar situation. For example, user may also want to ignore `KeyboardInput` when do preedit. We can't take into account all the cases I suppose.